File: coinChange.py
def can_make_amount_infinite_coins(coins, K):
    """
    Determines if it's possible to make amount K using infinite supply of given coins
    Using Dynamic Programming
    Time Complexity: O(n*K) where n is number of coin types
    Space Complexity: O(K)
    """
    dp = [False] * (K + 1)
    dp[0] = True  # Base case - zero amount is always possible
    
    for i in range(1, K + 1):
        for coin in coins:
            if i >= coin:
                dp[i] = dp[i] or dp[i - coin]
    
    return dp[K]

def can_make_amount_finite_coins(coins, counts, K):
    """
    Determines if it's possible to make amount K using finite supply of coins
    Using Dynamic Programming with coin counts
    Time Complexity: O(n*K*max(counts)) where n is number of coin types
    Space Complexity: O(K)
    """
    dp = [False] * (K + 1)
    dp[0] = True
    
    for coin, count in zip(coins, counts):
        for amount in range(K, -1, -1):
            for used in range(1, count + 1):
                if amount >= used * coin:
                    dp[amount] = dp[amount] or dp[amount - used * coin]
    
    return dp[K]

def min_coins_infinite_supply(coins, K):
    """
    Finds minimum number of coins needed to make amount K with infinite supply
    Using Dynamic Programming
    Time Complexity: O(n*K) where n is number of coin types
    Space Complexity: O(K)
    """
    dp = [float('inf')] * (K + 1)
    dp[0] = 0
    
    for i in range(1, K + 1):
        for coin in coins:
            if i >= coin:
                dp[i] = min(dp[i], dp[i - coin] + 1)
    
    return dp[K] if dp[K] != float('inf') else -1

def min_coins_finite_supply(coins, counts, K):
    """
    Finds minimum number of coins needed to make amount K with finite supply
    Using Dynamic Programming with coin counts
    Time Complexity: O(n*K*max(counts)) where n is number of coin types
    Space Complexity: O(K)
    """
    dp = [float('inf')] * (K + 1)
    dp[0] = 0
    
    for coin, count in zip(coins, counts):
        for amount in range(K, -1, -1):
            for used in range(1, count + 1):
                if amount >= used * coin:
                    if dp[amount - used * coin] != float('inf'):
                        dp[amount] = min(dp[amount], dp[amount - used * coin] + used)
    
    return dp[K] if dp[K] != float('inf') else -1

# Example usage:
if __name__ == "__main__":
    coins = [1, 2, 5]  # Coin denominations
    K = 11  # Target amount
    counts = [3, 2, 1]  # Number of each coin available
    
    # Test all scenarios
    print(f"Can make {K} with infinite coins:", 
          can_make_amount_infinite_coins(coins, K))
    
    print(f"Can make {K} with finite coins:", 
          can_make_amount_finite_coins(coins, counts, K))
    
    print(f"Minimum coins needed (infinite supply):", 
          min_coins_infinite_supply(coins, K))
    
    print(f"Minimum coins needed (finite supply):", 
          min_coins_finite_supply(coins, counts, K))

File: coinChange.cpp
#include <vector>
#include <climits>
#include <iostream>
using namespace std;

class CoinChange {
public:
    // Check if possible to make amount K with infinite supply of coins
    bool canMakeAmountInfiniteCoins(vector<int>& coins, int K) {
        vector<bool> dp(K + 1, false);
        dp[0] = true;  // Base case
        
        for(int i = 1; i <= K; i++) {
            for(int coin : coins) {
                if(i >= coin) {
                    dp[i] = dp[i] || dp[i - coin];
                }
            }
        }
        return dp[K];
    }
    
    // Check if possible to make amount K with finite supply of coins
    bool canMakeAmountFiniteCoins(vector<int>& coins, vector<int>& counts, int K) {
        vector<bool> dp(K + 1, false);
        dp[0] = true;
        
        for(int i = 0; i < coins.size(); i++) {
            for(int amount = K; amount >= 0; amount--) {
                for(int used = 1; used <= counts[i]; used++) {
                    if(amount >= used * coins[i]) {
                        dp[amount] = dp[amount] || dp[amount - used * coins[i]];
                    }
                }
            }
        }
        return dp[K];
    }
    
    // Find minimum coins needed with infinite supply
    int minCoinsInfiniteSupply(vector<int>& coins, int K) {
        vector<int> dp(K + 1, INT_MAX);
        dp[0] = 0;
        
        for(int i = 1; i <= K; i++) {
            for(int coin : coins) {
                if(i >= coin && dp[i - coin] != INT_MAX) {
                    dp[i] = min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[K] == INT_MAX ? -1 : dp[K];
    }
    
    // Find minimum coins needed with finite supply
    int minCoinsFiniteSupply(vector<int>& coins, vector<int>& counts, int K) {
        vector<int> dp(K + 1, INT_MAX);
        dp[0] = 0;
        
        for(int i = 0; i < coins.size(); i++) {
            for(int amount = K; amount >= 0; amount--) {
                for(int used = 1; used <= counts[i]; used++) {
                    if(amount >= used * coins[i] && dp[amount - used * coins[i]] != INT_MAX) {
                        dp[amount] = min(dp[amount], dp[amount - used * coins[i]] + used);
                    }
                }
            }
        }
        return dp[K] == INT_MAX ? -1 : dp[K];
    }
};

int main() {
    // Example usage
    vector<int> coins = {1, 2, 5};  // Coin denominations
    vector<int> counts = {3, 2, 1}; // Number of each coin available
    int K = 11;  // Target amount
    
    CoinChange solver;
    
    // Test all scenarios
    cout << "Can make " << K << " with infinite coins: " 
         << (solver.canMakeAmountInfiniteCoins(coins, K) ? "Yes" : "No") << endl;
    
    cout << "Can make " << K << " with finite coins: " 
         << (solver.canMakeAmountFiniteCoins(coins, counts, K) ? "Yes" : "No") << endl;
    
    cout << "Minimum coins needed (infinite supply): " 
         << solver.minCoinsInfiniteSupply(coins, K) << endl;
    
    cout << "Minimum coins needed (finite supply): " 
         << solver.minCoinsFiniteSupply(coins, counts, K) << endl;
    
    return 0;
}